### PR TITLE
fix: emit_serialized_unit_graph uses the configured shell

### PR DESCRIPTION
**Pull request by aatifsyed**
I'm writing a compatibility layer between the cargo and meson build systems.
This change allows me to plumb that through better :)

### CODE REVIEW

**Comment by epage** [comment]
What difference is this making for you?

**Comment by aatifsyed** [comment]
I'm using cargo-as-a-library to get the unit graph.

Before this change:
```
fn main() -> anyhow::Result<()> {
    match env::var_os("CARGO2MESON_CHILD").is_some() {
        // we're the child process, so run cargo so that it emits the unit graph
        true => {
            // ...
            let mut options = CompileOptions::new(&gctx, CompileMode::Build)?;
            options.build_config.unit_graph = true;
            cargo::ops::compile(&Workspace::new(&manifest_path, &gctx)?, &options)?;
        },
        // we're the parent process, so spawn ourselves and capture stdout
        false => {
            // ...
            let output = Command::new(env::current_exe()).env("CARGO2MESON_CHILD", "1") /* .arg(..) */ .output()?;
            serde_json::from_slice(&output.stdout)?;
            // ...
        }
    }
    Ok(())
}
```

After this change:
```
fn main() -> anyhow::Result<()> {
    // ...
    let shared_writer = SharedWriter::new(vec![]);
    let gtcx = GlobalContext::new(Shell::from_writer(Box::new(shared_writer.clone())), path, path);
    let mut options = CompileOptions::new(&gctx, CompileMode::Build)?;
    options.build_config.unit_graph = true;
    cargo::ops::compile(&Workspace::new(&manifest_path, &gctx)?, &options)?;
    serde_json::from_slice(&shared_writer.take())?;
    // ...
}
```

I won't have to workaround by spawning myself, and can use existing cargo mechanisms to graph the output :)

Happy to provide less abridged examples, but I hope the above gets the point across :)

File: src/cargo/core/compiler/unit_graph.rs [hunk 1]
```diff
@@ -127,9 +126,9 @@ pub fn emit_serialized_unit_graph(
...
     };
 
-    let stdout = std::io::stdout();
-    let mut lock = stdout.lock();
-    serde_json::to_writer(&mut lock, &s)?;
-    drop(writeln!(lock));
+    let mut shell = gctx.shell();
+    let stdout = shell.out();
+    serde_json::to_writer(&mut *stdout, &s)?;
+    writeln!(stdout)?;
```
Comment (weihanglo) [inline src/cargo/core/compiler/unit_graph.rs:132]: Maybe better using `gctx.shell().print_json(&result)?` as we have that function

**Review by weihanglo — DISMISSED** [review]
The change itself looks reasonable.

Just remind that libcargo is provided as-is, so you may find it painful when a `cargo` version bump is needed 😆.

File: src/cargo/core/compiler/unit_graph.rs (hunk 1 above)
Comment (aatifsyed) [reply src/cargo/core/compiler/unit_graph.rs]: Thanks! applied :)

**Comment by aatifsyed** [comment]
I've applied @weihanglo great suggestion :)

**Review by weihanglo — APPROVED** [review]

**Pull request merged**